pkg/sql/backfill.go

```diff
@@ -1728,55 +1845,146 @@ func ValidateForwardIndexes(
 	// Close when table count is ready.
 	tableCountsReady := make(chan struct{})
 
+	// Removes the protected timestamp, if one was added when this
+	// function returns.
+	var removeProtectedTS func(ctx context.Context) error
+	defer func() {
+		if removeProtectedTS != nil {
+			removeErr := removeProtectedTS(ctx)
+			if removeErr != nil {
+				if err != nil {
+					err = errors.WithSecondaryError(err, removeErr)
+				} else {
+					err = removeErr
+				}
+			}
+		}
+	}()
+	// Adds a protected timestamp only for historical queries.
+	var addProtectedTS = func() error {
+		var err error
+		protectedTSAdded.Do(func() {
+			// If we are not running a historical query, nothing to do here.
+			if runHistoricalTxn.ReadAsOf().IsEmpty() {
+				return
+			}
+			protectedtsID := uuid.MakeV4()
+			target := ptpb.MakeSchemaObjectsTarget(descpb.IDs{tableDesc.GetID()})
+			if err != nil {
+				return
+			}
+			rec := jobsprotectedts.MakeRecord(protectedtsID,
+				int64(jobID), runHistoricalTxn.ReadAsOf(), nil, jobsprotectedts.Jobs, target)
+			err = db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+				return protectedTSProvider.Protect(ctx, txn, rec)
+			})
+			if err != nil {
+				return
+			}
+			removeProtectedTS = func(ctx context.Context) error {
+				return db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+					return protectedTSProvider.Release(ctx, txn, protectedtsID)
+				})
+			}
```

Comment:
I know it isn't much code, but perhaps we could pull out a function here `protectTableForHistoricalTxn(ctx context.Context, runHistoricalTxn sqlutil.HistoricalInternalExecTxnRunner, table catalog.TableDescriptor) (func(ctx) error, error)`